src/components/Video.h

```diff
@@ -25,15 +25,41 @@ along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 
 #include <SDL_opengl.h>
 
+class VideoContext
+{
+public:
+  virtual void swapBuffers() = 0;
+};
+
+template<typename SwapBuffersF>
+class VideoContextImpl: public VideoContext
+{
+public:
+  VideoContextImpl(SwapBuffersF swapBuffers) : _swapBuffers(swapBuffers) {}
+  void swapBuffers() override { _swapBuffers(); }
+
+protected:
+  SwapBuffersF _swapBuffers;
+};
+
+template<typename SwapBuffersF>
+auto makeVideoContext(SwapBuffersF swapBuffers)
+{
+  return std::make_unique<VideoContextImpl<SwapBuffersF>>(swapBuffers);
```

Comment:
`std::make_unique` is c++14. the command line build currently targets c++11. 

I'm not sure if there's a requirement to not move to c++14. As a side note, c++14 is the lowest value we can have Visual Studio enforce: <https://docs.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=vs-2019>

Also note that the command line build in master is currently broken (I've fixed it in #179 (https://github.com/RetroAchievements/RALibretro/pull/179/commits/02d28d9992bb8061b97add67acf5436a2bdfd103))

Reply:
I can try to upgrade the command line build to c++14 (or both to c++17 for that matter). If there are blockers I can change this to be c++11 compliant.

Reply:
I'm not expecting you to change anything right now as it won't work without the changes that I referenced. I'll clean it up later. I'm more curious if @leiradel has any reason to keep it c++ 11.

Reply:
I don't, it's just the `VideoContext` thing that feels overkill to me. AFAIK it could be a virtual method of the video component, and all the data it needs to work would be properties of the video component implementation.

Reply:
Makes sense to me.  It seems straightforward enough to just pass and store the function pointer instead of wrapping it in a Context object. I'm not familiar with the RetroArch implementation, so I can't comment on the similiarities being drawn against that.

Reply:
Like I mentioned in https://github.com/RetroAchievements/RALibretro/pull/180#discussion_r432842170 a simple function pointer requires a new global variable to store the window handle, which doesn't sound better to me. There are other options, however. If we disregard the extensibility argument, I can simply pass the window handle to `Video::init` and use it without the need for callbacks, since the handle does not change during the lifetime of the component.

Reply:
That's what I was thinking in my previous comment. The video component is the place to store all things video related, so it seems to me like the best place to hold the window handle.

It's just that using classes or `std::function`s to implement components is a departure from the components with virtual methods, and I as a developer would find odd that some things are implemented one way, and some other in a different way.

Reply:
`init` is not a virtual function, so you can extend it however you want to make it behave appropriately without impacting the base interface. I think the disconnect here is that `Video` is generically named. Maybe it should be `SDLVideo` or something similar?

Reply:
That's a good point. `Video` is already tied to opengl, but moving the window handle in will tie it to SDL as well. Which is fine for now since it's the only supported combo, but from an architectural standpoint these should probably be two separate concerns. This is actually what retroarch does with video and context drivers, and what the `VideoContext` achieves, though I hadn't thought of it this way.

Maybe the `VideoContext` could be implemented as a component though, so that it fits better with the architectural style of RALibretro. Like the video component, it'd currently have a single implementation tied to SDL and opengl.

I'm fine with the simple solution of just putting the handle in `Video` for now though, if you prefer.

Reply:
Replaced the ad-hoc class with a `VideoContextComponent`.